contrib/mesos/pkg/scheduler/podtask/resources.go

```diff
@@ -0,0 +1,201 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package podtask
+
+import (
+	"math"
+
+	"github.com/gogo/protobuf/proto"
+	mesos "github.com/mesos/mesos-go/mesosproto"
+	"github.com/mesos/mesos-go/mesosutil"
+)
+
+func procureScalarResources(name string, value float64, roles []string, offer *mesos.Offer) []*mesos.Resource {
+	roleCaps := map[string]float64{}
+	res := filterResources(offer.Resources, isScalar, hasName(name))
+
+	for _, r := range res {
+		role := starred(r.GetRole())
+		roleCaps[role] += r.GetScalar().GetValue()
+	}
+
+	resources := []*mesos.Resource{}
+	for _, r := range roles {
+		if capacity := roleCaps[r]; capacity > 0 {
+			consumed := math.Min(capacity, value)
+			value -= consumed
+
+			res := mesosutil.NewScalarResource(name, consumed)
+			res.Role = protoNilString(r)
+			resources = append(resources, res)
+
+			if value == 0.0 {
+				break
+			}
+		}
+	}
+
+	if value > 0.0 {
+		return nil
+	}
+
+	return resources
+}
+
+// create a range resource for the listed ports
+func portRangeResources(spec *Spec) []*mesos.Resource {
+	rolePorts := make(map[string][]uint64)
+
+	for _, p := range spec.Ports {
+		rolePorts[p.Role] = append(rolePorts[p.Role], p.Port)
+	}
+
+	resources := []*mesos.Resource{}
+	for role, ports := range rolePorts {
+		resources = append(
+			resources,
+			&mesos.Resource{
+				Name:   proto.String("ports"),
+				Type:   mesos.Value_RANGES.Enum(),
+				Ranges: newRanges(ports),
+				Role:   protoNilString(role),
+			},
+		)
+	}
+
+	return resources
+}
+
+// generate port ranges from a list of ports. this implementation is very naive
+func newRanges(ports []uint64) *mesos.Value_Ranges {
+	r := make([]*mesos.Value_Range, 0)
+	for _, port := range ports {
+		x := proto.Uint64(port)
+		r = append(r, &mesos.Value_Range{Begin: x, End: x})
+	}
+	return &mesos.Value_Ranges{Range: r}
+}
+
+func foreachPortsRange(rs []*mesos.Resource, roles []string, f func(begin, end uint64, role string)) {
+	rs = filterResources(rs, hasName("ports"))
+	rs = byRoles(roles...).sort(rs)
+
+	for _, resource := range rs {
+		for _, r := range (*resource).GetRanges().Range {
+			bp := r.GetBegin()
+			ep := r.GetEnd()
+			f(bp, ep, (*resource).GetRole())
+		}
+	}
+}
+
+// byRolesSorter sorts resources according to the ordering of roles
+type byRolesSorter struct {
+	roles []string
+}
+
+// byRoles returns a byRolesSorter with the given roles
+func byRoles(roles ...string) *byRolesSorter {
+	return &byRolesSorter{roles: roles}
+}
+
+// sort sorts the given resources according to the order of roles in the byRolesSorter
+// and returns the sorted resources
+func (sorter *byRolesSorter) sort(resources []*mesos.Resource) []*mesos.Resource {
+	rolesMap := map[string][]*mesos.Resource{} // maps roles to resources
+	for _, res := range resources {
+		var roleResources []*mesos.Resource
+		role := starred(res.GetRole())
+
+		if _, ok := rolesMap[role]; !ok {
```

Comment:
Compare https://play.golang.org/p/O11HWFc4pa. Append know how to append to a nil list.